### PR TITLE
Find references to research for a given product

### PR DESCRIPTION
**Pull request by jwq1**
Think.. Wikipedia's "References" section (https://en.wikipedia.org/wiki/Nichols_and_Shepard)

screen shot 2017-12-13 at 1 00 01 am

### CODE REVIEW

**Comment by jwq1** [comment]
This code fully functions now. I can merge when you approve.

**Comment by hdavidzhu** [comment]
@jwq1 you can merge whenever in the future! don't let me block you. I'll just give you feedback to help you improve your dev skills, but keep hacking away!

**Comment by hdavidzhu** [comment]
Another reason why it's a good idea to make small PRs, especially since they are non-blocking 😄

File: final-project/application.py [hunk 1]
```diff
@@ -486,6 +486,84 @@ def search_category():
...
+# Store references to product research
+@app.route("/reference")
+def reference():
+    """Search for references to product research."""
+
+    # Keep the user logged in
+    user_id = session.get("user_id")
+
+    # Get the product name requested by the users
+    product_to_provide_references_for = request.args.get(
```
Comment (hdavidzhu) [inline final-project/application.py:498]: not sure why you split this into 2 lines. Did it go over 80 chars? This looks less readable on Github.

File: final-project/application.py [hunk 2]
```diff
@@ -486,6 +486,84 @@ def search_category():
...
+        return apology("Please specify product")
+
+    # Search for references to a given product
+    references_for_product = db.execute("""
+            SELECT research.link, title
+            FROM research
+            LEFT JOIN products
+            ON products.id=research.product_id
+            WHERE products.product_name=:product
+            """, product=product_to_provide_references_for)
```
Comment (hdavidzhu) [inline final-project/application.py:514]: I like these new SQL commands!

File: final-project/application.py [hunk 3]
```diff
@@ -486,6 +486,84 @@ def search_category():
...
+        if not references_for_product[reference]["title"]:
+            # If there is no title, then use the link.
+            reference_titles.append(references_for_product[reference]["link"])
+        # If there is a title, then use the title.
+        else:
+            # Save the title.
+            reference_titles.append(references_for_product[reference]["title"])
+
+        # Save the link.
+        reference_links.append(references_for_product[reference]["link"])
```
Comment (hdavidzhu) [inline final-project/application.py:544]: This works fine, though I wonder if you also like to have titles and links associated w/ each other? In that case, I would look at tuples or other data structures to keep the data together.

Also take a look at http://book.pythontips.com/en/latest/map_filter.html to make your python more functional and declarative. They are easier to read, less state-dependent, and easier to debug! Woo for functional programming!

For future-future reference, something like this is also fun 😄  https://github.com/ReactiveX/RxPY

File: final-project/application.py [hunk 4]
```diff
@@ -486,6 +486,84 @@ def search_category():
...
+            , number_of_references=number_of_references
+            , reference_titles=reference_titles
+            , reference_links=reference_links
+            , product_to_provide_references_for=product_to_provide_references_for)
+
+    return apology("TODO")
+
+
+# TODO: Add references to product research
+@app.route("/add_reference")
```
Comment (hdavidzhu) [inline final-project/application.py:558]: Have you learned about the REST API format? I notice a lot of your routes are named liked methods. This isn't the mainstream convention, and I'd take some time to learn the standard.

I'm sure there's a good resource somewhere here: https://www.youtube.com/results?search_query=rest+api&page=&utm_source=opensearch

File: final-project/sqlite-schemas.txt [hunk 5]
```diff
@@ -0,0 +1,60 @@
+CREATE TRIGGER add_research_title AFTER INSERT ON research FOR EACH ROW BEGIN INSERT INTO research (title) VALUES (research.link); END;
```
Comment (hdavidzhu) [inline final-project/sqlite-schemas.txt]: Why is this checked in?

If it's for future documentation, might I recommend Markdown?

File: final-project/application.py (hunk 1 above)
Comment (jwq1) [reply final-project/application.py:498]: Yes, it was 82 characters or something. Where do you draw the line with number characters? Gut feeling about what is more or less readable?

File: final-project/application.py (hunk 3 above)
Comment (jwq1) [reply final-project/application.py:544]: The information is already in a list of dictionary objects, which makes it pretty easy to access. However, I wasn't sure if I should store the information into variables for readability. When do you create variables versus just keep everything in a more complex data structure?

For things like `request.args.get(something_from_the_get_request)`, would this information immediately get stored into a variable for reuse? Then for things like `reference[0]["title"]` and `reference[0]["link"]`, do I keep the information in the list of dictionary objects?

File: final-project/application.py (hunk 4 above)
Comment (jwq1) [reply final-project/application.py:558]: I remember seeing REST appear like wildfire on Hacker News a year back, but that is the extent of it. I was always curious. I will look more into it.

File: final-project/sqlite-schemas.txt (hunk 5 above)
Comment (jwq1) [reply final-project/sqlite-schemas.txt]: I never thought to stick it in my own files. I'll use markdown for documentation now. 

Looks good! https://guides.github.com/features/mastering-markdown/

Why this is checked in... umm.. laziness on my part? :

I kept having to write and re-write these SQL commands to debug the database. My plan was to delete this file when finished. However, maybe there is a better way to keep track of this.

Comment (jwq1) [reply final-project/sqlite-schemas.txt]: The other thing is, the SQLite databases are not that flexible, I cannot rename, remove, or edit data types in columns once they are created. If I want to make any major changes, I have to drop the table then repopulate it. The queries in this document are easy to paste in to repopulate the database with some initial products and research. 

Probably better for me to keep these on my local branch. Like you mentioned earlier, if I was working with other developers I would never check-in a file like this.

Comment (hdavidzhu) [reply final-project/sqlite-schemas.txt]: Hmmm, seems like you need migrations for those cases, not re-creations of new DBs

**Pull request merged**